### ISSUE TITLE
apache-fakegooglebot jail is broken in Ubuntu 16.04.1 LTS

### ISSUE THREAD

**Issue by maksyms** [fail2ban/fail2ban#1506]
### Environment:
- Fail2Ban version 0.9.3
- Ubuntu 16.04.1 LTS
- [X] Fail2Ban installed via OS/distribution mechanisms
- [X] You have not applied any additional foreign patches to the codebase
- [ ] Some customizations were done to the configuration (provide details below is so)

### The issue:

Enabling apache-fakegooglebot jail in Ubuntu 16.04.1 LTS leads to all Google bots to be banned.

#### Steps to reproduce

Enable apache-fakegooglebot jail and reload configuration.

#### Expected behavior

Google bots not blocked. malicious IPs pretending to be google bots blocked.

#### Observed behavior

All google bots blocked.

#### Any additional information
The root cause of the issue is that fail2ban 0.9.3 installs its fail2ban.server.filter package into Python 3. On my Ubuntu 16.04.1 LTS, which was an upgrade from Ubuntu 14.04 LTS, Python 2.7 is the default. This leads to /etc/fail2ban/filter.d/ignorecommands/apache-fakegooglebot to fail when looking for DNSUtils in fail2ban.server.filter.

The solution is trivial: change "#!/usr/bin/python" in /etc/fail2ban/filter.d/ignorecommands/apache-fakegooglebot to "#!/usr/bin/python3". Then everything works as expected.

I will try to find time to update this soon.

#### Relevant parts of /var/log/fail2ban.log file:

```
2016-08-10 11:41:37,053 fail2ban.filter         [3381]: INFO    [apache-fakegooglebot] Found 66.249.76.50
2016-08-10 11:41:37,271 fail2ban.action         [3381]: ERROR   /etc/fail2ban/filter.d/ignorecommands/apache-fakegooglebot 66.249.76.50 -- stdout: b''
2016-08-10 11:41:37,271 fail2ban.action         [3381]: ERROR   /etc/fail2ban/filter.d/ignorecommands/apache-fakegooglebot 66.249.76.50 -- stderr: b'Traceback (most recent call last):\n  File "/etc/fail2ban/filter.d/ignorecommands/apache-fakegooglebot", line 32, in <module>\n    is_googlebot(process_args(sys.argv))\n  File "/etc/fail2ban/filter.d/ignorecommands/apache-fakegooglebot", line 17, in process_args\n    from fail2ban.server.filter import DNSUtils\nImportError: No module named fail2ban.server.filter\n'
2016-08-10 11:41:37,272 fail2ban.action         [3381]: ERROR   /etc/fail2ban/filter.d/ignorecommands/apache-fakegooglebot 66.249.76.50 -- returned 1
```

**Comment by sebres** [comment]
> The root cause of the issue is that fail2ban 0.9.3 installs its fail2ban.server.filter package into Python 3.

Nope, I suspect that you has a hodge-podge with defaults python, at least since (or even before) the fail2ban was installed.

Facts are:
- fail2ban should be always installed using default python bindings;
- default python version should be `/usr/bin/python`, that normally a link to /usr/bin/python2 (or /usr/bin/python3)
- if you switch your default python version (do it correct), you should definitely reinstall fail2ban.

BTW. If for any reason whatsoever, the /usr/bin/python is not the default version, the changing to `#!/usr/bin/env python` would be more safe variant...
Seems to be already "fixed" so in 0.10 (see https://github.com/fail2ban/fail2ban/blob/0.10/config/filter.d/ignorecommands/apache-fakegooglebot#L1)
But still not done in 0.9 (see https://github.com/fail2ban/fail2ban/blob/0.9/config/filter.d/ignorecommands/apache-fakegooglebot#L1)

Thus I'll close the issue, please reopen if I'm wrong.

**Comment by maksyms** [comment]
Interesting.

So why then fail2ban-0.9.3, which is the default on Ubuntu 16.04.1 LTS, has shebang as #!/usr/bin/python3?

Here's an exerpt from fail2ban-client:

**#!/usr/bin/python3**
# emacs: -*- mode: python; py-indent-offset: 4; indent-tabs-mode: t -*-
# vi: set ft=python sts=4 ts=4 sw=4 noet :

and here is the same from fail2ban-server:

#!/usr/bin/python3
# emacs: -*- mode: python; py-indent-offset: 4; indent-tabs-mode: t -*-
# vi: set ft=python sts=4 ts=4 sw=4 noet :

Then the whole package on debian is wrong, isn't it?

**Comment by sebres** [comment]
I'll control what does this...

**Comment by maksyms** [comment]
Also, I cannot reopen the issue, as you are the repo collaborator and I'm not. I still think there's an issue as the Ubuntu version of fail2ban has python3 hard-coded into it.

**Comment by maksyms** [comment]
Thank you!

**Comment by sebres** [comment]
Thx you.
Well, as long as #1426 was not back-ported to 0.9 branch (only 0.10), it may be theoretical a feature of `2to3` (which runs during setup)... Or a resultant problem of executing that, if Ubuntu fail2ban installer runs not in default python version (precisely because `2to3` can modify the python files, which will be not compatible to the default version). 

I'll cherry pick it to master (0.9) branch, if we've clarified that...

The first question would be - if all `bin/fail2ban-*` files are modified so, why the same not occurred with another "executable" pythonic files (like apache-fakegooglebot)...
Second one would be - why and whether Ubuntu installed it not with default python version.
Then it could be an Ubuntu issue...

Can you please execute following test on your system (and give here the output):
```bash
python  -c  "import sys, subprocess; print(sys.version); print('------\n'); subprocess.call(['/usr/bin/env', 'python', '--version'])"
python2 -c  "import sys, subprocess; print(sys.version); print('------\n'); subprocess.call(['/usr/bin/env', 'python', '--version'])"
python3 -c  "import sys, subprocess; print(sys.version); print('------\n'); subprocess.call(['/usr/bin/env', 'python', '--version'])"
```

**Comment by maksyms** [comment]
Sure, here you go:

```
[maksym@ksusha 14:35:08 ~]$ python  -c  "import sys, subprocess; print(sys.version); print('------\n'); subprocess.call(['/usr/bin/env', 'python', '--version'])"
2.7.12 (default, Jul  1 2016, 15:12:24)
[GCC 5.4.0 20160609]
------

Python 2.7.12
[maksym@ksusha 14:35:09 ~]$ python2 -c  "import sys, subprocess; print(sys.version); print('------\n'); subprocess.call(['/usr/bin/env', 'python', '--version'])"
2.7.12 (default, Jul  1 2016, 15:12:24)
[GCC 5.4.0 20160609]
------

Python 2.7.12
[maksym@ksusha 14:35:09 ~]$ python3 -c  "import sys, subprocess; print(sys.version); print('------\n'); subprocess.call(['/usr/bin/env', 'python', '--version'])"
3.5.2 (default, Jul  5 2016, 12:43:10)
[GCC 5.4.0 20160609]
------

Python 2.7.12
[maksym@ksusha 14:35:33 ~]$
```

**Comment by sebres** [comment]
Fact is your default python is 2.7.

Are you sure, that you have not switched the python default version to 2.7 after installing of fail2ban?

**Comment by maksyms** [comment]
I know it is. Yes, I am absolutely sure I haven't switched the default version of Python. Fail2ban is installed on my systems for the last 3-4 years or so.

Switching the default version also doesn't explain how fail2ban-server and fail2ban-client have explicit python3 mentioned.,,

I've been reading and I can see that /usr/bin/python is not supposed to be python3 on Ubuntu. [Here is the link](https://wiki.ubuntu.com/Python/3)

**Comment by maksyms** [comment]
And here is another [link](http://askubuntu.com/questions/777085/why-isnt-python-3-the-default-python-binary-in-ubuntu-16-04) stating the same.

**Comment by sebres** [comment]
I know that, but the question is if the default python 2.7, why and by what the fail2ban was installed with python3...
BTW. The question is rather rhetorical (makes possibly sometimes sense). I'll see how we could fix the execution of pythonic actions with the same version as the fail2ban running...

**Comment by maksyms** [comment]
Thanks, sebres.

**Comment by mikini** [comment]
For reference;
This is still an issue on a fresh Ubuntu 16.04.2 LTS.
Ubuntu bug reported as [#1708197](https://bugs.launchpad.net/ubuntu/+source/fail2ban/+bug/1708197).

**Comment by sebres** [comment]
@mikini 
Currently fixed in 0.10, debian/0.9.7-1 ... 0.9.6 only (38d53a72fd2b55fae0c6a57e7ea10ffa3446b650), for older versions you can workaround this using full-path to python-version, that used for fail2ban (for which fail2ban was installed), like here (example for python 2.7):
```diff
-#!/usr/bin/env python
+#!/usr/bin/python2.7
```
[Here a link](https://github.com/fail2ban/fail2ban/commit/a4cef8a893317d3204fa97aaa3b7a4136dc6df2e#diff-0c5bfb7a0c02686bdb0b12f3458f2439R1) for the fix (Note that command `fail2ban-python` does not exists in older versions).

If you've it already, please check the version returned by:
```bash
fail2ban-python -V
```
And compare with version of python, that you see by fail2ban daemon (use `top`, `ps` etc).

**Comment by mikini** [comment]
Thanks for your time looking at this oldie. I had devised the same workaround (as indicated in the Ubuntu report). The comment was primarily meant as a heads up to fail2ban maintainers (hi there) that this is still happening on the current and widely used Ubuntu LTS and as an indication for other Ubuntu users falling into the same trap, that the issue is known and reported in Ubuntu.

For your reference here are full specifications of the setup on an up to date Ubuntu 16.04.2 LTS server:
```
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.2 LTS
Release:	16.04
Codename:	xenial
```
fail2ban is 0.9.3:
```
$ apt-cache policy fail2ban
fail2ban:
  Installed: 0.9.3-1
  Candidate: 0.9.3-1
  Version table:
 *** 0.9.3-1 500
        500 http://dk.archive.ubuntu.com/ubuntu xenial/universe amd64 Packages
        500 http://dk.archive.ubuntu.com/ubuntu xenial/universe i386 Packages
        100 /var/lib/dpkg/status
```
Default python being 2:
```
$ python -V && python2 -V && python3 -V
Python 2.7.12
Python 2.7.12
Python 3.5.2
```

fail2ban-server running under python3:
```
$ systemctl status fail2ban.service 
● fail2ban.service - Fail2Ban Service
   Loaded: loaded (/lib/systemd/system/fail2ban.service; enabled; vendor preset: enabled)
   Active: active (running) since Tue 2017-08-01 12:55:22 UTC; 6 days ago
     Docs: man:fail2ban(1)
 Main PID: 1565 (fail2ban-server)
    Tasks: 23
   Memory: 47.7M
      CPU: 13min 53.629s
   CGroup: /system.slice/fail2ban.service
           └─1565 /usr/bin/python3 /usr/bin/fail2ban-server -s /var/run/fail2ban/fail2ban.sock -p /var/run/fail2ban/fail2ban.pid -x -b

Aug 01 12:55:21 gc-web01 systemd[1]: Starting Fail2Ban Service...
Aug 01 12:55:21 gc-web01 fail2ban-client[1241]: 2017-08-01 12:55:21,814 fail2ban.server         [1537]: INFO    Starting Fail2ban v0.9.3
Aug 01 12:55:21 gc-web01 fail2ban-client[1241]: 2017-08-01 12:55:21,814 fail2ban.server         [1537]: INFO    Starting in daemon mode
Aug 01 12:55:22 gc-web01 systemd[1]: Started Fail2Ban Service.

```

Searching all shebangs of the fail2ban package shows that python3 is hardcoded into the main fail2ban scripts, with my edit of apache-fakegooglebot visible (python used in test scripts, which shouldn't be a production issue):
```
$ dpkg -L fail2ban|xargs -i bash -c 'if [ -f {} ]; then if out=$(grep -F "#!" {}); then echo -e {}:\\n$out; fi; fi'
/usr/share/doc/fail2ban/examples/ipmasq-ZZZzzz_fail2ban.rul:
#! /bin/sh
/usr/share/doc/fail2ban/examples/cacti/fail2ban_stats.sh:
#!/bin/bash
/usr/share/doc/fail2ban/FILTERS.gz:
Binary file /usr/share/doc/fail2ban/FILTERS.gz matches
/usr/share/man/man1/jail.conf.10.gz:
Binary file /usr/share/man/man1/jail.conf.10.gz matches
/usr/bin/fail2ban-regex:
#!/usr/bin/python3
/usr/bin/fail2ban-client:
#!/usr/bin/python3
/usr/bin/fail2ban-testcases:
#!/usr/bin/python3
/usr/bin/fail2ban-server:
#!/usr/bin/python3
/usr/lib/python3/dist-packages/fail2ban/tests/files/config/apache-auth/digest.py:
#!/usr/bin/python
/usr/lib/python3/dist-packages/fail2ban/tests/files/ignorecommand.py:
#!/usr/bin/python
/usr/lib/python3/dist-packages/fail2ban/tests/actiontestcase.py:
 f.write("""#!/bin/bash
/etc/fail2ban/filter.d/ignorecommands/apache-fakegooglebot:
#!/usr/bin/python3
/etc/init.d/fail2ban:
#! /bin/sh
```
